src/qt3utils/applications/controllers/princeton_spectrometer.py

```diff
@@ -0,0 +1,142 @@
+import numpy as np
+import tkinter as tk
+import logging
+from typing import Tuple
+
+import qt3utils.datagenerators.princeton as princeton
+
+
+class QT3ScanPrincetonSpectrometerController:
+    """
+    Implements qt3utils.applications.qt3scan.interface.QT3ScanSpectrometerDAQControllerInterface
+    """
+    def __init__(self, logger_level: int):
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(logger_level)
+
+        self.spectrometer = princeton.SpectrometerDataAcquisition()
+        self.last_config_dict = {}
+
+        self.wave_start = None
+        self.wave_end = None
+        self.last_measured_spectrum = None
+        self.last_wavelength_array = None
+
+    @property
+    def clock_rate(self) -> float:
+        try:
+            _t = self.spectrometer.exposure_time / 1000.0  # Converting from milliseconds to seconds.
+        except Exception as e:
+            self.logger.error(e)
+            _t = 2  # TODO: better default behavior. Should this be -1? 1? or should Spectrometer be changed.
+        return 1.0 / _t
+
+    def start(self) -> None:
+        """
+        Nothing to be done in this method. All acquisition is happening in the "sample_spectrum" method.
+        """
+        self.logger.debug('calling QT3ScanPrincetonSpectrometerController start')
```

Comment:
Then why is it here? Or why not implement the sample_spectrum in here?

Reply:
@vasilisniaouris  there's some detritus left-over from my initial pass at refactoring last fall. The definition of the interface requires this method.  It could be better named to 'initialize_before_acquisition' perhaps. Or something. Depending on the long-term plans for this code, this would be a good cleanup task -- go through the interfaces and qt3scan, understand what's happening and then rename methods appropriately.